xdsl/dialects/vector.py

```diff
@@ -383,6 +397,330 @@ def __init__(
+            )
+        if seen[expr.position]:
+            raise VerifyException(
+                f'"{op.name}" requires a permutation_map that is a permutation (found one dim used more than once)'
+            )
+        seen[expr.position] = True
+
+
+def verify_transfer_op(
+    op: TransferReadOp | TransferWriteOp,
+    shaped_type: MemRefType[Attribute] | TensorType[Attribute],
+    vector_type: VectorType[Attribute],
+    mask_type: VectorType[I1] | None,
+    # WJOG9GVF: TODO fix: remove None type from inferred_mask_type once 7S4F0FZA has been fixed
+    inferred_mask_type: VectorType[I1] | None,
+    permutation_map: AffineMap,
+    in_bounds: ArrayAttr[BoolAttr],
+):
+    """
+    This mirrors VectorOps.cpp -> verifyTransferOp from MLIR
+    """
+
+    element_type = shaped_type.element_type
+    vector_element_type = vector_type.element_type
+
+    if isa(element_type, VectorType[Attribute]):
+        # Memref or tensor has vector element type
+        # TODO verify vector element type
+        pass
+    else:
+        # Memref of tensor has scalar element type
+        if isa(vector_element_type, IndexType):
+            if not isa(element_type, IndexType):
+                raise VerifyException(
+                    "Element type of source is index, expected element type of vector also to be index"
+                )
+        else:
+            assert isa(vector_element_type, IntegerType | AnyFloat)
+            assert isa(element_type, IntegerType | AnyFloat)
+
+            minor_size = (
+                1 if vector_type.get_num_dims() == 0 else vector_type.get_shape()[-1]
+            )
+            result_vec_size = vector_element_type.bitwidth * minor_size
+            if result_vec_size % element_type.bitwidth != 0:
+                raise VerifyException(
+                    f'"{op.name}" requires the bitwidth of the minor 1-D vector to be an integral multiple of the bitwidth of the source element type'
+                )
+
+        # Check that permutation map results match rank of vector type.
+        if len(permutation_map.results) != vector_type.get_num_dims():
+            raise VerifyException(
+                f'"{op.name}" requires a permutation_map with result dims of the same rank as the vector type'
+            )
+
+    if permutation_map.num_symbols != 0:
+        raise VerifyException(f'"{op.name}" requires permutation_map without symbols')
+
+    if permutation_map.num_dims != shaped_type.get_num_dims():
+        raise VerifyException(
+            f'"{op.name}" requires a permutation_map with input dims of the same rank as the source type'
+        )
+
+    # WJOG9GVF: TODO fix: uncomment this when 7S4F0FZA has been fixed
+    # if mask_type:
+    #     if mask_type != inferred_mask_type:
+    #         raise VerifyException(
+    #             f'"{op.name}" inferred mask type ({inferred_mask_type}) and mask operand type ({mask_type}) don\'t match'
+    #         )
```

Comment:
?

Reply:
Same as below. I will move it to a new pull request